pynipap/tracing.py

```diff
@@ -0,0 +1,110 @@
+    from opentelemetry.sdk.trace.export import BatchSpanProcessor
+    import opentelemetry.exporter.otlp.proto.http.trace_exporter
+
+    tracer = trace.get_tracer("pynipap")
+
+
+    def init_tracing(service_name, endpoint, use_grpc=True):
+        resource = Resource(attributes={
+            SERVICE_NAME: service_name
+        })
+
+        provider = TracerProvider(resource=resource)
+        if use_grpc:
+            processor = BatchSpanProcessor(OTLPSpanExporter(endpoint=endpoint))
+        else:
+            processor = BatchSpanProcessor(opentelemetry.exporter.otlp.proto.http.trace_exporter.OTLPSpanExporter(endpoint=endpoint))
+        provider.add_span_processor(processor)
+        trace.set_tracer_provider(provider)
+
+
+    def create_span(f):
+        """ Class decorator creating a opentelemetry span
+        """
+
+        @wraps(f)
+        def decorated(*args, **kwargs):
+            """
+            """
+
+            signature = inspect.signature(f)
+
+            if args[0].__class__ == type:
+                span_name = args[0].__name__ + " " + f.__name__
+            else:
+                span_name = str(args[0].__class__.__name__) + " " + f.__name__
+
+            with tracer.start_as_current_span(span_name,
+                                              context.get_current(),
+                                              SpanKind.CLIENT):
+                index = 0
+                span = trace.get_current_span()
+                for parameter in signature.parameters:
+                    if index > 0:
+                        try:
+                            span.set_attribute("argument." + parameter, str(args[index]))
+                        except IndexError:
+                            break
+                    index += 1
+                try:
+                    span.set_attribute("username", AuthOptions().options['username'])
+                    span.set_attribute("authoritative_source", AuthOptions().options['authoritative_source'])
+                except BaseException:
+                    pass
+                return f(*args, **kwargs)
+        return decorated
+
+
+    class TracingXMLTransport(xmlrpclib.Transport):
+
+        def __init__(self, use_datetime=False, use_builtin_types=False,
+                     *, headers=()):
+            super().__init__(use_datetime=use_datetime,
+                             use_builtin_types=use_builtin_types,
+                             headers=headers)
+
+        def request(self, host, handler, request_body, verbose=False):
+            with tracer.start_as_current_span("POST XML request", context.get_current(), SpanKind.CLIENT):
+                current_span = trace.get_current_span()
+                try:
+                    result = super().request(host, handler, request_body, verbose)
+                    current_span.set_attribute("http.status_code", 200)
+                except xmlrpclib.ProtocolError as protocolError:
```

Comment:
Shouldn't we make sure to re-raise the exception, this'll silently drop all ProtocolErrors?

Reply:
Yes, we definitely should. I've pushed an update correcting this.